src/transformStateWithClones.js

```diff
@@ -7,7 +7,34 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let copyState = { ...state };
+  const result = [];
+
+  actions.forEach((item) => {
+    const type = item.type;
+    const extra = item.extraData;
+    const keys = item.keysToRemove;
+
+    switch (type) {
+      case 'addProperties':
+        copyState = { ...copyState, ...extra };
+        break;
+
+      case 'removeProperties':
+        keys.forEach((key) => {
+          delete copyState[key];
```

Comment:
Directly deleting properties from `copyState` can lead to unintended side effects. Consider creating a new object without the specified keys to ensure immutability. For example, use `copyState = Object.keys(copyState).reduce((acc, key) => { if (!keys.includes(key)) acc[key] = copyState[key]; return acc; }, {});`.